grpc/src/main/java/io/stargate/grpc/service/Service.java

```diff
@@ -94,38 +134,47 @@ private void handleError(Throwable throwable, StreamObserver<Result> responseObs
     if (throwable instanceof StatusException || throwable instanceof StatusRuntimeException) {
       responseObserver.onError(throwable);
     } else {
-      responseObserver.onError(Status.UNKNOWN.withCause(throwable).asRuntimeException());
+      responseObserver.onError(
+          Status.UNKNOWN
+              .withDescription(throwable.getMessage())
+              .withCause(throwable)
+              .asRuntimeException());
     }
   }
 
-  private void prepare(
-      Connection connection, Query query, BiConsumer<Prepared, ? super Throwable> afterPrepare) {
-    final StringBuilder keyBuilder = new StringBuilder(query.getCql());
+  private CompletableFuture<Prepared> prepareQuery(
+      Connection connection, String cql, StringValue keyspace, boolean tracing) {
+    CompletableFuture<Prepared> future = new CompletableFuture<>();
+    final StringBuilder keyBuilder = new StringBuilder();
     connection.loggedUser().ifPresent(user -> keyBuilder.append(user.name()));
-    QueryParameters queryParameters = query.getParameters();
-    if (queryParameters.getKeyspace().isInitialized()) {
-      keyBuilder.append(queryParameters.getKeyspace().getValue());
+    if (keyspace.isInitialized()) {
+      keyBuilder.append(keyspace.getValue());
     }
+    keyBuilder.append(cql);
     final String key = keyBuilder.toString();
     // Caching here to avoid round trip to the persistence backend thread.
     Prepared prepared = preparedCache.getIfPresent(key);
     if (prepared != null) {
-      afterPrepare.accept(prepared, null);
+      future.complete(prepared);
     } else {
+      ImmutableParameters.Builder parameterBuilder =
+          ImmutableParameters.builder().tracingRequested(tracing);
+      if (keyspace.isInitialized()) {
```

Comment:
If it's not initialized then there's is no keyspace set for the query use. Instead, it might be in the query string instead e.g. "SELECT keyspace.table ...".